src/transformStateWithClones.js

```diff
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  const clone = { ...state };
+
+  for (const action of actions) {
+    const { type, extraData, keysToRemove } = action;
+
+    switch (type) {
+      case 'addProperties':
+        for (const key in extraData) {
+          clone[key] = extraData[key];
+        }
+        break;
+
+      case 'removeProperties':
+        for (const key of keysToRemove) {
+          delete clone[key];
+        }
+        break;
+
+      case 'clear':
+        for (const key in clone) {
+          delete clone[key];
+        }
+        break;
+    }
+
+    result.push({ ...clone });
+  }
+
+  return result;
```

Comment:
The current implementation mutates the `clone` object across different actions, which violates the requirement to not mutate objects. Instead, you should create a new copy of the state inside each case of the switch statement.